In [2]:
import pandas as pd
import numpy as np

## 1. Attendance Data

### Load the attendance.csv file

In [3]:
att = pd.read_csv('untidy-data/attendance.csv')

In [4]:
att.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [13]:
att.shape

(4, 9)

In [14]:
att.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        4 non-null      object
 1   2018-01-01  4 non-null      object
 2   2018-01-02  4 non-null      object
 3   2018-01-03  4 non-null      object
 4   2018-01-04  4 non-null      object
 5   2018-01-05  4 non-null      object
 6   2018-01-06  4 non-null      object
 7   2018-01-07  4 non-null      object
 8   2018-01-08  4 non-null      object
dtypes: object(9)
memory usage: 416.0+ bytes


In [12]:
# rename columns 

att.rename(columns={'Unnamed: 0':'name'}, inplace=True)
att.head()

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [17]:
att_melt = att.melt(id_vars=['name'], var_name='date', value_name='attendance')
att_melt.head()

name        date attendance
0  Sally  2018-01-01          P
1   Jane  2018-01-01          A
2  Billy  2018-01-01          A
3   John  2018-01-01          P
4  Sally  2018-01-02          T

### Calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [18]:
att_melt.attendance.value_counts()

T    14
P     9
A     6
H     3
Name: attendance, dtype: int64

In [26]:
att_melt.name.value_counts()

John     8
Jane     8
Billy    8
Sally    8
Name: name, dtype: int64

In [27]:
# change attendance to numeric

def convert_att(x):
    if x == 'P':
        return 1
    elif x == 'H':
        return 0.5
    elif x == 'T':
        return 0.9
    else:
        return 0
    
att_melt['att_value'] = att_melt['attendance'].apply(convert_att)
att_melt.head()

name        date attendance  att_value
0  Sally  2018-01-01          P        1.0
1   Jane  2018-01-01          A        0.0
2  Billy  2018-01-01          A        0.0
3   John  2018-01-01          P        1.0
4  Sally  2018-01-02          T        0.9

In [39]:
# percentage of each student's attendance = sum of student attendance/ number of days for attendance

# sum of each student's attendance
att_sum = att_melt.groupby('name').att_value.sum()
att_sum

name
Billy    4.2
Jane     5.5
John     7.3
Sally    6.1
Name: att_value, dtype: float64

In [41]:
# number of days for attendance
total_days = att_melt.date.unique().size
total_days
# or do this: att_melt.date.value_counts().count()

8

In [42]:
# percentage of each student's attendance
att_sum / total_days

name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: att_value, dtype: float64

## 2. Coffee Levels

### a. Read the coffee_levels.csv file.

In [43]:
coffee = pd.read_csv('untidy-data/coffee_levels.csv')

In [44]:
coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

### b. Transform the data so that each carafe is in it's own column.

In [45]:
coffee.coffee_carafe.value_counts()

z    10
y    10
x    10
Name: coffee_carafe, dtype: int64

In [46]:
c_long = coffee.pivot_table(index = 'hour',
                  columns = 'coffee_carafe', values = 'coffee_amount')

In [47]:
c_long

coffee_carafe         x         y         z
hour                                       
8              0.816164  0.189297  0.999264
9              0.451018  0.521502  0.915990
10             0.843279  0.023163  0.144928
11             0.335533  0.235529  0.311495
12             0.898291  0.017009  0.771947
13             0.310711  0.997464  0.398520
14             0.507288  0.058361  0.864464
15             0.215043  0.144644  0.436364
16             0.183891  0.544676  0.280621
17             0.391560  0.594126  0.436677

### c. Is this the best shape for the data?

In [48]:
coffee.shape, c_long.shape

((30, 3), (10, 3))

- yes, because it is in tidy data format

## 3. Cake Recipes

### a. Read the cake_recipes.csv data

In [60]:
c_recipes = pd.read_csv('untidy-data/cake_recipes.csv')

In [61]:
c_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [69]:
c_recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   225       8 non-null      float64
 1   250       8 non-null      float64
 2   275       8 non-null      float64
 3   300       8 non-null      float64
 4   recipe    8 non-null      object 
 5   position  8 non-null      object 
dtypes: float64(4), object(2)
memory usage: 512.0+ bytes


In [62]:
c_recipes.shape

(8, 5)

### Note:
- This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

### b. Tidy the data as necessary

### Initial Thoughts
- make separate columns for recipe and oven rack position 
- drop recipe:position column
- separate column for cake tastiness score

In [63]:
# split recipe:position col

c_recipes[['recipe', 'position']] = c_recipes['recipe:position'].str.split(':', expand = True)
c_recipes.head()

recipe:position        225        250        275        300 recipe position
0        a:bottom  61.738655  53.912627  74.414730  98.786784      a   bottom
1           a:top  51.709751  52.009735  68.576858  50.228470      a      top
2        b:bottom  57.095320  61.904369  61.196980  99.248541      b   bottom
3           b:top  82.455004  95.224151  98.594881  58.169349      b      top
4        c:bottom  96.470207  52.001358  92.893227  65.473084      c   bottom

In [64]:
c_recipes = c_recipes.drop(columns='recipe:position')
c_recipes.head()

225        250        275        300 recipe position
0  61.738655  53.912627  74.414730  98.786784      a   bottom
1  51.709751  52.009735  68.576858  50.228470      a      top
2  57.095320  61.904369  61.196980  99.248541      b   bottom
3  82.455004  95.224151  98.594881  58.169349      b      top
4  96.470207  52.001358  92.893227  65.473084      c   bottom

In [78]:
# melt table for separate tastiness score and oven temp columns

cr_melt = c_recipes.melt(id_vars = ['recipe','position'], 
        var_name = 'oven_temp',
      value_name = 'tastiness_score')
cr_melt.head()

recipe position oven_temp  tastiness_score
0      a   bottom       225        61.738655
1      a      top       225        51.709751
2      b   bottom       225        57.095320
3      b      top       225        82.455004
4      c   bottom       225        96.470207

### c. Which recipe, on average, is the best?

In [67]:
# find highest avg score
cr_melt.groupby('recipe').tastiness_score.mean().sort_values(ascending=False)

recipe
b    76.736074
c    75.874748
a    63.922201
d    62.864844
Name: tastiness_score, dtype: float64

- Recipe b has the highest average tastiness score so it is the best recipe.

### d. Which oven temperature, on average, produces the best results?

In [80]:
# find oven temp with highest avg score 
cr_melt.groupby('oven_temp').tastiness_score.mean()

oven_temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: tastiness_score, dtype: float64

- Since 275 oven temp has the highest avg tatsiness score we can say that temp produces the best results

### e. Which combination of recipe, rack position, and temperature gives the best result? 

In [93]:
cr_melt.groupby(['oven_temp', 'position', 'recipe']).tastiness_score.mean().nlargest(1)

oven_temp  position  recipe
300        bottom    b         99.248541
Name: tastiness_score, dtype: float64

In [90]:
cr_melt.groupby(['oven_temp', 'recipe', 'position']).tastiness_score.mean().idxmax()

('300', 'b', 'bottom')

In [97]:
# highest tastiness score
cr_melt.tastiness_score.max()

99.2485405378462